django/db/models/sql/compiler.py

```diff
@@ -1403,22 +1404,25 @@ def as_sql(self):
                     )
             else:
                 val = field.get_db_prep_save(val, connection=self.connection)
-
+            key_lookups = self.query.key_lookups.get(field.column)
             # Getting the placeholder for the field.
             if hasattr(field, 'get_placeholder'):
-                placeholder = field.get_placeholder(val, self, self.connection)
+                placeholder = field.get_placeholder(val, self, self.connection, key_lookups=key_lookups)
             else:
                 placeholder = '%s'
-            name = field.column
+
+            quoted_name = qn(field.column)
+            sql_placeholder = field.get_sql_placeholder(quoted_name, key_lookups)
+
             if hasattr(val, 'as_sql'):
                 sql, params = self.compile(val)
-                values.append('%s = %s' % (qn(name), placeholder % sql))
+                values.append(sql_placeholder % (placeholder % sql))
                 update_params.extend(params)
             elif val is not None:
-                values.append('%s = %s' % (qn(name), placeholder))
+                values.append(sql_placeholder % placeholder)
                 update_params.append(val)
             else:
-                values.append('%s = NULL' % qn(name))
+                values.append(sql_placeholder % 'NULL')
```

Comment:
It feels like the new or adjusted API should call a new `as_update_sql(lhs)` method on the lhs that passes the rhs and expect a `(str: sql, Tuple[Any]: params)` value to be returned. We have to make sure that by default this method raises by default so transforms like `update(date__year=2020)` fail but they could likely be implemented.

That would require `UpdateQuery.add_update_values` to solve the lookups and resolve both lhs and rhs while disallowing joins and thus aliasing. You can achieve that by sure making `UpdateQuery.alias_cols` is always `False`.